microsoft-edge/progressive-web-apps-chromium/how-to/widgets.md

```diff
@@ -0,0 +1,495 @@
+    // The widget just got installed, render it using renderWidget.
+    // Pass the event.widget object to the function.
+    event.waitUntil(renderWidget(event.widget));
+  }
+});
+
+async function renderWidget(widget) {
+  // Get the template and data URLs from the widget definition.
+  const templateUrl = widget.definition.msAcTemplate;
+  const dataUrl = widget.definition.data;
+
+  // Fetch the template text and data.
+  const template = await (await fetch(templateUrl)).text();
+  const data = await (await fetch(dataUrl)).text();
+
+  // Render the widget with the template and data.
+  await self.widgets.updateByTag(widget.definition.tag, {template, data});
+}
+```
+
+#### Update widgets on service worker updates
+
+When the service worker code changes in a PWA, the browser detects that change, installs the new service worker, and then later activates it.
+
+When this happens, it is important to update any widget instances that may already be running. Widgets may have been installed before the service worker `activate` event is emitted. Update your widgets when this event occurs to avoid displaying empty widgets.
+
+```javascript
+// Update the widgets to their initial states
+// when the service worker is activated.
+self.addEventListener("activate", event => {
+  event.waitUntil(updateWidgets());
+});
+
+async function updateWidgets() {
+  // Get the widget that match the tag defined in the web app manifest.
+  const widget = await self.widgets.getByTag("pwamp");
+  if (!widget) {
+    return;
+  }
+
+  // Using the widget definition, get the template and data.
+  const template = await (await fetch(widget.definition.msAcTemplate)).text();
+  const data = await (await fetch(widget.definition.data)).text();
+
+  // Render the widget with the template and data.
+  await self.widgets.updateByTag(widget.definition.tag, {template, data});
+}
+```
+
+#### Handle widget actions
+
+If the widget template contains actions, users can execute those actions by clicking buttons in the rendered widget. See [Define widget actions](#define-widget-actions) for information about how to define actions in a template.
+
+When a user executes a widget action, a `widgetclick` event is triggered in the PWA service worker. To handle the user action, listen to the event:
+
+```javascript
+self.addEventListener('widgetclick', (event) => {
+  switch (event.action) {
+    case 'previous-song':
+      // Application logic to play the previous song...
+      break;
+    case 'next-song':
+      // Application logic to play the next song...
+      break;
+  }
+});
+```
+
+For brevity, the actual application code is not shown in the above code snippet. When the `previous-song` or `next-song` actions are received, a message would probably need to be sent to the app using [Client.postMessage](https://developer.mozilla.org/docs/Web/API/Client/postMessage) to let the app know that it should start playing the previous or next songs.
```

Comment:
```suggestion
For brevity, the actual application code isn't shown in the above code snippet. When the `previous-song` or `next-song` actions are received, a message would probably need to be sent to the app using [Client.postMessage](https://developer.mozilla.org/docs/Web/API/Client/postMessage) to let the app know that it should start playing the previous or next songs.
```
Acrolinx would probably suggest contraction, per style guide (style guide is mainly about end-user docs not Dev docs).  Sometimes I avoid contraction in API Ref docs for Dev audience where I find that it would change the meaning & weaken communication.  Case-by-case.